<launch>
    <param name="/use_sim_time" value="true"/>

    <arg name="task_file" />

    <arg name="nogui_flag" default=""/>
    <arg name="reallocate" default="false"/>

    <arg name="dynamic_mechanism" default="false" />

    <arg name="dummy_robot_name" default="robot_0"/>

    <arg name="classifier_name" default="clf_execution_phase_time_random_forest"/>

    <arg name="model" default="burger" doc="model type [burger, waffle, waffle_pi]"/>

    <!-- Map server -->
    <node name="map_server" pkg="map_server" type="map_server" args="$(find mrta)/config/maps/$(arg map_file)"
          respawn="false"/>

    <!-- Stage -->
    <node pkg="stage_ros" type="stageros" name="stageros"
          args="$(arg nogui_flag) $(find mrta)/config/stage/$(arg world_file)" required="true" respawn="false">
        <param name="base_watchdog_timeout" value="0.2"/>
    </node>

    <!--&lt;!&ndash; BEGIN Dummy Robot &ndash;&gt;-->
    <!--<group ns="$(arg dummy_robot_name)">-->
        <!--<param name="tf_prefix" value="$(arg dummy_robot_name)"/>-->

        <!--&lt;!&ndash; turtlebot3_kcl_navigation/move_base.launch &ndash;&gt;-->
        <!--<include file="$(find turtlebot3_kcl_navigation)/launch/move_base.launch">-->
            <!--<arg name="model" value="$(arg model)" />-->
        <!--</include>-->

        <!--<node pkg="fake_localization" type="fake_localization" name="fake_localization" respawn="false" output="screen">-->
            <!--<param name="odom_frame_id" value="$(arg dummy_robot_name)/odom"/>-->
            <!--<param name="base_frame_id" value="$(arg dummy_robot_name)/base_link"/>-->
            <!--<param name="publish_frequency" value="5.0"/>-->
        <!--</node>-->
    <!--</group>-->
    <!--&lt;!&ndash; END Dummy Robot &ndash;&gt;-->

    <!-- BEGIN ROBOT 1 -->
    <group ns="robot_1">
        <param name="tf_prefix" value="robot_1"/>

        <!--&lt;!&ndash; turtlebot3_kcl_navigation/amcl.launch &ndash;&gt;-->
        <!--<include file="$(find turtlebot3_kcl_navigation)/launch/amcl.launch">-->
            <!--<arg name="odom_frame_id" value="robot_1/odom"/>-->
            <!--<arg name="base_frame_id" value="robot_1/base_link"/>-->
            <!--<arg name="scan_topic" value="robot_1/base_scan"/>-->
        <!--</include>-->

        <!-- turtlebot3_kcl_navigation/move_base.launch -->
        <include file="$(find turtlebot3_kcl_navigation)/launch/move_base.launch">
            <arg name="model" value="$(arg model)" />
            <arg name="cmd_vel_topic" value="robot_1/cmd_vel"/>
            <arg name="odom_topic" value="robot_1/odom"/>
            <remap from="scan" to="base_scan"/>
        </include>

        <node pkg="fake_localization" type="fake_localization" name="fake_localization" respawn="false" output="screen">
            <param name="odom_frame_id" value="robot_1/odom"/>
            <param name="base_frame_id" value="robot_1/base_link"/>
            <param name="publish_frequency" value="5.0"/>
        </node>

        <!--
        Arguments to mrta_robot_controller:
          robot_name    - e.g. "robot_1"
          reallocate    - True/False: are we doing task reallocation?
          is_turtlebot  - True/False
        -->
        <node name="mrta_robot_controller" pkg="mrta_robot_controller" type="mrta_robot_controller" respawn="false"
              required="true" output="screen">
            <param name="robot_name" type="str" value="robot_1"/>
            <param name="reallocate" type="bool" value="$(arg reallocate)"/>
            <param name="is_turtlebot" type="bool" value="False"/>
        </node>

        <!--&lt;!&ndash;  ***************** Robot Model *****************  &ndash;&gt;-->
        <!--<include file="$(find turtlebot3_bringup)/launch/includes/description.launch.xml">-->
            <!--<arg name="model" value="$(arg model)" />-->
        <!--</include>-->

        <!--<node pkg="joint_state_publisher" type="joint_state_publisher" name="joint_state_publisher">-->
            <!--<param name="use_gui" value="false"/>-->
            <!--<param name="rate" value="50"/>-->
        <!--</node>-->

        <!--<node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher">-->
            <!--<param name="publish_frequency" type="double" value="50.0" />-->
            <!--<param name="tf_prefix" type="string" value="robot_1" />-->
        <!--</node>-->

    </group>
    <!-- END ROBOT 1 -->

    <!-- BEGIN ROBOT 2 -->
    <group ns="robot_2">
        <param name="tf_prefix" value="robot_2"/>

        <!-- turtlebot3_kcl_navigation/move_base.launch -->
        <include file="$(find turtlebot3_kcl_navigation)/launch/move_base.launch">
            <arg name="model" value="$(arg model)" />
            <arg name="cmd_vel_topic" value="robot_2/cmd_vel"/>
            <arg name="odom_topic" value="robot_2/odom"/>
        </include>

        <node pkg="fake_localization" type="fake_localization" name="fake_localization" respawn="false" output="screen">
            <param name="odom_frame_id" value="robot_2/odom"/>
            <param name="base_frame_id" value="robot_2/base_link"/>
            <param name="publish_frequency" value="5.0"/>
        </node>

        <!--
        Arguments to mrta_robot_controller:
          robot_name    - e.g. "robot_1"
          reallocate    - True/False: are we doing task reallocation?
          is_turtlebot  - True/False
        -->
        <node name="mrta_robot_controller" pkg="mrta_robot_controller" type="mrta_robot_controller" respawn="false"
              required="true" output="screen">
            <param name="robot_name" type="str" value="robot_2"/>
            <param name="reallocate" type="bool" value="$(arg reallocate)"/>
            <param name="is_turtlebot" type="bool" value="False"/>
        </node>

        <!--&lt;!&ndash;  ***************** Robot Model *****************  &ndash;&gt;-->
        <!--<include file="$(find turtlebot3_bringup)/launch/includes/description.launch.xml">-->
            <!--<arg name="model" value="$(arg model)" />-->
        <!--</include>-->

        <!--<node pkg="joint_state_publisher" type="joint_state_publisher" name="joint_state_publisher">-->
            <!--<param name="use_gui" value="false"/>-->
            <!--<param name="rate" value="50"/>-->
        <!--</node>-->

        <!--<node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher">-->
            <!--<param name="publish_frequency" type="double" value="50.0" />-->
        <!--</node>-->

    </group>
    <!-- END ROBOT 2 -->

    <!-- BEGIN ROBOT 3 -->
    <group ns="robot_3">
        <param name="tf_prefix" value="robot_3"/>

        <!-- turtlebot3_kcl_navigation/move_base.launch -->
        <include file="$(find turtlebot3_kcl_navigation)/launch/move_base.launch">
            <arg name="model" value="$(arg model)" />
            <arg name="cmd_vel_topic" value="robot_3/cmd_vel"/>
            <arg name="odom_topic" value="robot_3/odom"/>
        </include>

        <node pkg="fake_localization" type="fake_localization" name="fake_localization" respawn="false" output="screen">
            <param name="odom_frame_id" value="robot_3/odom"/>
            <param name="base_frame_id" value="robot_3/base_link"/>
            <param name="publish_frequency" value="5.0"/>
        </node>

        <!--
        Arguments to mrta_robot_controller:
          robot_name    - e.g. "robot_1"
          reallocate    - True/False: are we doing task reallocation?
          is_turtlebot  - True/False
        -->
        <node name="mrta_robot_controller" pkg="mrta_robot_controller" type="mrta_robot_controller" respawn="false"
              required="true" output="screen">
            <param name="robot_name" type="str" value="robot_3"/>
            <param name="reallocate" type="bool" value="$(arg reallocate)"/>
            <param name="is_turtlebot" type="bool" value="False"/>
        </node>

        <!--&lt;!&ndash;  ***************** Robot Model *****************  &ndash;&gt;-->
        <!--<include file="$(find turtlebot3_bringup)/launch/includes/description.launch.xml">-->
            <!--<arg name="model" value="$(arg model)" />-->
        <!--</include>-->

        <!--<node pkg="joint_state_publisher" type="joint_state_publisher" name="joint_state_publisher">-->
            <!--<param name="use_gui" value="false"/>-->
            <!--<param name="rate" value="50"/>-->
        <!--</node>-->

        <!--<node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher">-->
            <!--<param name="publish_frequency" type="double" value="50.0" />-->
        <!--</node>-->

    </group>
    <!-- END ROBOT 3 -->

    <node name="mrta_auctioneer" pkg="mrta_auctioneer" type="mrta_auctioneer" required="true" output="screen">
        <param name="task_file" value="$(find mrta_auctioneer)/task_files/$(arg task_file)"/>
        <param name="scenario_id" value="$(arg scenario_id)"/>
        <param name="mechanism" value="$(arg mechanism)"/>
        <param name="reallocate" value="$(arg reallocate)"/>
        <param name="dynamic_mechanism" value="$(arg dynamic_mechanism)" />
        <param name="dummy_robot_name" value="$(arg dummy_robot_name)"/>
        <param name="classifier_name" value="$(arg classifier_name)"/>
    </node>

</launch>
